feat: add flyout wc spec (#3691)
microsoft/fast: specs/flyout/flyout.md

### Background
A Flyout can be less intrusive to the user experience than a Dialog and should be used instead of a Dialog when information or input is needed to be displayed, but not in the center of the screen, such as in the case of collection of additional info before an action, displaying information that is only relevant some of the time, or displaying more information.

### Use Cases
- As a Dropdown for a menu, navigation bar, etc.
- In a context menu as the secondary menu/action.
- As a popup showing additional info for an item on the page.
- To display a larger image of a thumbnail image in a photo gallery.
- To display an action related to a page item

### Features
**Placement Logic:** - This component will use *Anchored Region* to be positioned relative to the Flyouts anchor element. This way it can be dynamic and adapt to the available space if `responsive` is true for the Flyout.

**Light Dismiss (aka Soft Dismiss):** - A way to quickly close the Flyout by clicking outside it, or pressing the `esc` key.

> **Overlay:** - A control of whether or not the Flyout has an overlay effect that dims the surrounding UI when the Flyout is active.

Reviewer: How common is this with other instances of flyout-like controls across libraries? This seems more common with a dialog pattern, whereas things like flyout, callout, coachmarks seem to rarely have overlays. If this is common, can you link to some examples here?

Reviewer 2: Just browsing around for new features, but I have an example I find quite common - it'd be for tutorial/intro/demo interactions where a user is guided to perform actions, sometimes not allowed to proceed without completing required action.

### Prior Art/Examples
##### Context Menu
*Note:* the "Manage, Apply, Show more options" is the Flyout for this example.

> ![Context Menu](https://docs.microsoft.com/en-us/windows/uwp/design/controls-and-patterns/images/flyout-nested.png)

Reviewer: I actually don't think this would be a flyout here - I think this would be menu/menu item which already exists. Perhaps it might be better to list these examples by Design system rather than "type". I would expect that the base flyout is primarily going to be similar to dialog or where it is a container with semantics and behaviors. It would likely be helpful to have additional examples from other design systems as currently it seems like we only have UWP examples. An example from Fluent might be their "callout" or "teaching bubble". Other design systems may call this "popover". I think having other examples will help here.

##### Text Only
*Note:* the button is not part of the Flyout, but shown for context as an anchor element.
![Text Only](https://docs.microsoft.com/en-us/windows/uwp/design/controls-and-patterns/images/flyout-wrapping-text.png)

##### Actionable
*Note:* the button is not part of the Flyout, but shown for context as an anchor element.
![Actionable More Info](https://docs.microsoft.com/en-us/windows/uwp/design/controls-and-patterns/images/flyout-example2.png)

---

### API
*Component Name*

> - `fast-flyout`

Reviewer: There seem to be quite a few names related to this type of control; while I've referred to it as "flyout" it seems like a more common term is "popover". If we're ultimately hoping to be aligned to open ui, we should probably look for a name that has the best path forward for alignment.

*Attributes*

> - `anchor` - string, the html id of the HTMLElement that the Flyout is attached to, (viewport for anchored region defaults to the parent of the anchor)

Reviewer: I know that in "anchored region" the attribute is "anchor", but I wonder if that is too specific to a *type* of element. These are often triggered by elements with a role of button. I'm not sure I would expect "button" to be a replacement, but something more broad such as "trigger" seems more reasonable. What do other design systems do for this?

Author: Seems like target or trigger is used in others. I think target is good, but this will mean that we won't be using common names across our own components as tooltip uses _*anchor*_

Author: @chrisdholt I also think it may be beneficial to keep it as anchor as we may be able to reuse the tooltip control and only change a few things to get a popover to work. The content and trigger method will be different, as well as the role, but the control will be very similar.

Author: This was changed to target.

- `overlay` - boolean, specifies whether an overlay is drawn to dim the out of scope UI until the Flyout is dismissed

> - `light-dismiss` - boolean, specifies whether or not the flyout can be dismissed by clicking elsewhere or pressing `esc` key, defaults to true

Reviewer 3: I sometimes see this called "soft dismiss" - did you happen to look at other libraries to see what is more common?

Reviewer: I think there's been some discussion or usage of terminology over in Open UI. Might be good to dig into issues and PR's to see how they've referred to it. In Dialog, we emit an event regardless in the case of light dismiss rather than having at attribute here. Would that be preferred here as well? I think the benefit is that if you want to do something on dismiss you listen for the event; if you don't want to - you don't. Thoughts on creating alignment between these two?

- `visible` - boolean, whether or not the Flyout is visible, defaults to undefined, as this is controlled by the author

> - `responsive` - boolean, whether or not the positioning is responsive based on available space, defaults to true, uses Anchored Region logic

Reviewer 4: Works for me.  We should try to have similar "position selection" options for this, tooltip and other components that have this kind of behavior.

# Popover (aka Flyout)
## Overview

> Popover is a focusable floating container that shows over a pages content and displays UI related to what the user is doing and is typically triggered by a clickable element. Similar to a Dialog, but with positioning controls, no overlay option, and no specific action is required, a Popover can be used to reveal a secondary control or show more detail about an item.

Reviewer 5: ```suggestion
Popover is a focusable floating container that shows over a page's content and displays UI related to what the user is doing and is typically triggered by a clickable element. Similar to a Dialog, but with positioning controls, no overlay option, and no specific action is required. A Popover can be used to reveal a secondary control or show more detail about an item.
```

### Background

> A Popover can be less intrusive to the user experience than a Dialog and should be used instead of a Dialog when information or input is needed to be displayed, but not in the center of the screen, such as in the case of collection of additional info before an action, displaying information that is only relevant some of the time, or just displaying more information.

Reviewer 5: ```suggestion
A Popover can be less intrusive to the user experience than a Dialog and should be used instead of a Dialog when information or input is needed to be displayed, but not in the center of the screen, such as in the case of the collection of additional info before an action, displaying information that is only relevant some of the time or for just displaying more information.
```

### Use Cases
- As a popup showing additional info for an item on the page.
- To display a larger image of a thumbnail image in a photo gallery.
- To display an action related to a page item.

### Features
**Placement Logic:** - This component will use *Anchored Region* to be positioned relative to the Popover anchor element. This way it can be dynamic and adapt to the available space.

> **Soft Dismiss (aka Light Dismiss):** - A way to quickly close the Popover by clicking outside it, or pressing the `esc` key. An event would be emitted anytime the Popover is dismissed. The Popover would always be soft-dismissable. Detecting the users next click of a different element can trigger this when clicking outside the Popover.

Reviewer 5: ```suggestion
**Soft Dismiss (aka Light Dismiss):** - A way to quickly close the Popover by clicking outside it, or pressing the `esc` key. An event would be emitted anytime the Popover is dismissed. The Popover would always be soft-dismissable. Detecting the user's next click of a different element can trigger this when clicking outside the Popover.
```

> **Focus Management:** - Should follow the same process for focus management as a Dialog primarily as the Popover's role will be dialog. This includes a trap focus method, which can be optional. For accessibility the focus should return to the target that triggers the popover after the popover is closed.

Reviewer 5: ```suggestion
**Focus Management:** - Should follow the same process for focus management as a Dialog primarily as the Popover's role will be dialog. This includes a trap focus method, which can be optional. For accessibility, the focus should return to the target that triggers the popover after the popover is closed.
```

### Prior Art/Examples

##### Popover from Salesforce
- https://www.lightningdesignsystem.com/components/popovers/

![Popover](./images/salesforce-popover.png)
##### Popover from Evergreen

> - triggered by a button

Reviewer 5: ```suggestion
- Triggered by a button
```

- https://evergreen.segment.com/components/popover/

![Popover](./images/evergreen-popover.png)

##### Microsoft Windows Developer Flyout, Text Only

> *Note:* the button is not part of the Popover, but shown for context as an anchor element.

Reviewer 5: ```suggestion
*Note:* The button is not part of the Popover, but shown for context as an anchor element.
```

![Text Only](https://docs.microsoft.com/en-us/windows/uwp/design/controls-and-patterns/images/flyout-wrapping-text.png)

##### Microsoft Windows Developer Flyout, Actionable

> *Note:* the button is not part of the Popover, but shown for context as an anchor element.

Reviewer 5: ```suggestion
*Note:* The button is not part of the Popover, but shown for context as an anchor element.
```

**ASK (2)**
```html
${when(
    x => x.visible,
    html<Popover>`
        <div>
            <fast-anchored-region
                anchor="${x => x.targetElement}"
                vertical-positioning-mode="dynamic"
                horizontal-positioning-mode="dynamic"
                horizontal-default-position="${x => x.getHorizontalPosition()}"
                vertical-default-position="${x => x.getVerticalPosition()}"
                horizontal-inset="${x => x.getHorizontalInset()}"
                vertical-inset="${x => x.getVerticalInset()}"
                ${ref("anchoredRegion")}
            >
                <div
                    class="content"
                    part="content"
                    role="dialog"
                >
                    <slot></slot>
                </div>
            </fast-anchored-region>
        </div>
    `
)}
```

---

## Implementation
```html
<fast-popover visible>
    <p>This is a popover</p>
    <fast-button>Action</fast-button>
</fast-popover>
```

### States
`visible` - Whether or not Popover is hidden. This state is managed solely by the app author via the visible attribute.

### Accessibility
*Keyboard Navigation and Focus*

> Keyboard and navigation will follow the same rules as Dialog per the [W3C Specification](https://w3c.github.io/aria-practices/#dialog_modal), expect when there is no focusable content then the Popover itself will receive focus.

Reviewer 5: ```suggestion
Keyboard and navigation will follow the same rules as Dialog per the [W3C Specification](https://w3c.github.io/aria-practices/#dialog_modal), except when there is no focusable content then the Popover itself will receive focus.
```

### Globalization

> The component visuals should change when in RTL scenarios as the component is positioned relative to it's anchor.

Reviewer 5: ```suggestion
The component visuals should change when in RTL scenarios as the component is positioned relative to its anchor.
```

# Popover (aka Flyout)
## Overview

> Popover is a focusable floating container that shows over a pages content and displays UI related to what the user is doing. It is typically triggered by a clickable element. Similar to a Dialog, but with positioning controls, no overlay option, and no specific action being required, a Popover can be used to reveal a secondary control or show more detail about an item.

Reviewer 6: > with positioning controls

Are the positioning controls part of the UI or are they part of the API?

Author: They are part of the API

### Use Cases
- As a popup showing additional info for an item on the page.
- To display a larger image of a thumbnail image in a photo gallery.
- To display an action related to a page item.

### Features
**Placement Logic:** - This component will use *Anchored Region* to be positioned relative to the Popover target element. This way it can be dynamic and adapt to the available space.

**Soft Dismiss (aka Light Dismiss):** - A way to quickly close the Popover by clicking outside it, or pressing the `esc` key. An event would be emitted anytime the Popover is dismissed. The Popover would always be soft-dismissable. Detecting the users next click of a different element can trigger this when clicking outside the Popover.

**Focus Management:** - Should follow the same process for focus management as a Dialog, primarily as the Popover's role will be 'dialog'.

This includes:
- Trap focus - optional, defaults to enabled - Focus is contained within the popover while it is visible.
- First element focus - Focusing on the first focusable element when trap focus is enabled, but not overriding an element that is set to be focused by the author.

> - Returning focus to the target - optional, defaults to enabled, after popover is closed the focus would return to the target element.

Reviewer 3: Does the flyout close automatically on scroll?

> - Returning focus to the target - optional, defaults to enabled, after popover is closed the focus would return to the target element.

Author: No it does not. I never saw that behavior when looking at other flyout and popover components.

![Text Only](https://docs.microsoft.com/en-us/windows/uwp/design/controls-and-patterns/images/flyout-wrapping-text.png)

#### Microsoft Windows Developer Flyout, Actionable
![Actionable More Info](https://docs.microsoft.com/en-us/windows/uwp/design/controls-and-patterns/images/flyout-example2.png)

#### Inline Dialog from ATLAssian
- https://atlassian.design/components/inline-dialog/examples

![Popover](./images/atlassian-popover.png)

#### Callout with actions from Fluent UI
- https://developer.microsoft.com/en-us/fluentui#/controls/web/callout
![Callout](./images/fluent-callout.png)

---

### API
*Component Name*
- `fast-popover`

*Attributes*
- `visible` - boolean, whether or not the Popover is visible, defaults to undefined, as this is controlled by the author
- `trap-focus` - boolean, whether or not to keep focus contained inside the popover, defaults to true
- `target` - string, the html id of the HTMLElement that the Popover is attached to, (viewport for anchored region defaults to the parent of the anchor) and triggered by
- `focus-target-on-close` - boolean, whether or not to focus back on to the _target element_ when the popover is closed, defaults to true

> - `position` - enum, where the tooltip should appear relative to its target, uses Anchored Region logic=

Reviewer 6: This should describe all valid values for `position`, or link to the Anchored Region documentation for `position` (or both).

Reviewer: +1 - Should we support the same positioning values as tooltip? Less? More?

Author: I have updated with a note to see the _Position Options_ section of this spec, as it has several options available.

> - `delay` - number, the time delay before the popover is shown once triggered

Reviewer 6: What unit of time is the delay value?

Author: ms, updated

- `aria-labelledby` - optional based on implementation**
- `aria-describedby` - optional based on implementation**
- `aria-label` - optional, based on implementation**

    ** See the [W3C Specification](https://w3c.github.io/aria-practices/#dialog_roles_states_props) for requirements and details.

*Properties*

> - `targetElement` - the reference to the HTMLElement used as an target

Reviewer 6: ```suggestion
- `targetElement` - the reference to the HTMLElement used as a target
```

Author: updated

> - `viewportElement` - the reference to the HTMLElement used as the viewport, required ot setup anchored region properly

Reviewer 6: ```suggestion
- `viewportElement` - the reference to the HTMLElement used as the viewport, required to set up anchored region properly
```

Author: updated

> - `popoverVisible` - boolean, whether or not the Popover is visible

Reviewer 6: How about `isVisible` or `open`?

Author: In keeping with how Tooltip has this, I think staying with popoverVisible helps clearly separate it from the `visible` attr.

Author: updated to flyoutVisible

Reviewer: Is there a visible attr? If so, there is a visible JavaScript property. What would the difference be between a "visible" attribute and this property?

*Slots*
- default for content

*Events*
- `dismiss` - event fired when the Popover is dismissed

*Functions*

> - `createPopoverTemplate(string: prefix)` - generates a `ViewTemplate` for the popover based on the given provided prefix string. This is required as popover uses an `anchored-region` internally and the create function generates a template using the appropriate `anchored-region` tag (ie "fast-anchored-region", "fluent-anchored-region"). Note that the appropriate `anchored-region` component must also be declared.

Reviewer 6: `createTemplate` would make it easier to rename this component if we decide to go with something other than `popover`.

Author: We currently have 2 other create***Template functions that are public and exported so I think we need to have Popover in there for clarification.

*Position Options*

- above-centered, below-centered

![X](./images/ab-centered.png)

- right-centered, left-centered

![X](./images/rl-centered.png)

- top-left-corner, top-right-corner, bottom-right-corner, bottom-left-corner

![X](./images/corner-positions.png)

- above-left-aligned, below-left-aligned

![X](./images/ab-left-aligned.png)

- above-right-aligned, below-right-aligned

![X](./images/ab-right-aligned.png)

- left-top-aligned, right-top-aligned

![X](./images/rl-top-aligned.png)

- left-bottom-aligned, right-bottom-aligned

![X](./images/rl-bottom-aligned.png)

### Anatomy and Appearance
Parts:
- `popover` - the popover itself, has `role=dialog`
- `anchored-region` - the anchored region that controls the position of the popover

*Template:*
**ASK (2)**
```html
${when(
    x => x.visible,
    html<Popover>`
        <div>

>             <fast-anchored-region

Reviewer 6: I ran into this issue when building `fast-select` - I initially used other fast components in the template, which creates a weird circular dependency where we need fast-components to build fast-foundation, which is needed to build fast-components. This also means that the `fast-` tag "namespace" will need to be permanent, which could create conflicts for authors who want to compose their own components off of `fast-foundation`.

Author: Ahh yes, this will be resolved with a recent change similar to Tooltip where a prefix is given to createPopoverTemplate, but i failed to include that here in this example. Updated now.

> # Flyout (aka Popover)

Reviewer: We should drop "popover" for now as I believe we've moved away from that name.

### Use Cases
- As a popup showing additional info for an item on the page.
- To display a larger image of a thumbnail image in a photo gallery.
- To display an action related to a page item.

### Features
**Placement Logic:** - This component will use *Anchored Region* to be positioned relative to the Flyout target element. This way it can be dynamic and adapt to the available space.

**Soft Dismiss (aka Light Dismiss):** - A way to quickly close the Flyout by clicking outside it, or pressing the `esc` key. An event would be emitted anytime the Flyout is dismissed. The Flyout would always be soft-dismissable. Detecting the users next click of a different element can trigger this when clicking outside the Flyout.

**Focus Management:** - Should follow the same process for focus management as a Dialog, primarily as the Flyout's role will be 'dialog'.

This includes:

> - Trap focus - optional, defaults to enabled - Focus is contained within the Flyout while it is visible.

Reviewer: We should add this to the risks section - we're currently re-evaluating how we do this, but trapping focus w/r/t the Shadow DOM can be quite problematic: https://nolanlawson.com/2021/02/13/managing-focus-in-the-shadow-dom/

Author: Added!

### Use Cases
- As a popup showing additional info for an item on the page.
- To display a larger image of a thumbnail image in a photo gallery.
- To display an action related to a page item.

### Features
**Placement Logic:** - This component will use *Anchored Region* to be positioned relative to the Flyout target element. This way it can be dynamic and adapt to the available space.

**Soft Dismiss (aka Light Dismiss):** - A way to quickly close the Flyout by clicking outside it, or pressing the `esc` key. An event would be emitted anytime the Flyout is dismissed. The Flyout would always be soft-dismissable. Detecting the users next click of a different element can trigger this when clicking outside the Flyout.

**Focus Management:** - Should follow the same process for focus management as a Dialog, primarily as the Flyout's role will be 'dialog'.

This includes:
- Trap focus - optional, defaults to enabled - Focus is contained within the Flyout while it is visible.
- First element focus - Focusing on the first focusable element when trap focus is enabled, but not overriding an element that is set to be focused by the author.

> - Returning focus to the target - optional, defaults to enabled, after Flyout is closed the focus would return to the target element.

Author: recapturing original comments:
https://github.com/microsoft/fast/pull/3691#discussion_r554124652

- https://evergreen.segment.com/components/popover/

![Popover](./images/evergreen-popover.png)

#### Microsoft Windows Developer Flyout, Text Only
![Text Only](https://docs.microsoft.com/en-us/windows/uwp/design/controls-and-patterns/images/flyout-wrapping-text.png)

#### Microsoft Windows Developer Flyout, Actionable
![Actionable More Info](https://docs.microsoft.com/en-us/windows/uwp/design/controls-and-patterns/images/flyout-example2.png)

#### Inline Dialog from ATLAssian
- https://atlassian.design/components/inline-dialog/examples

![Popover](./images/atlassian-popover.png)

#### Callout with actions from Fluent UI
- https://developer.microsoft.com/en-us/fluentui#/controls/web/callout
![Callout](./images/fluent-callout.png)

---

### API
*Component Name*
- `fast-flyout`

*Attributes*
- `visible` - boolean, whether or not the flyout is visible, defaults to undefined, as this is controlled by the author
- `trap-focus` - boolean, whether or not to keep focus contained inside the Flyout, defaults to true
- `target` - string, the html id of the HTMLElement that the Flyout is attached to, (viewport for anchored region defaults to the parent of the anchor) and triggered by

> - `focus-target-on-close` - boolean, whether or not to focus back on to the _target element_ when the Flyout is closed, defaults to true

Author: This is definitely open to other naming ideas or approaches to solving this problem, which is described in the focus management section comments

Author: After re-reading the spec, it is expected that the target that triggers a dialog is focused on once the dialog is closed and since we have set the role of flyout to dialog we should follow that. It is currently defaulted to do so, but should we allow something like `focus-taget-on-close` or not even make that optional?

Reviewer 6: Maybe intercepting the `dismiss` event to redirect focus somewhere else (or prevent the default focusing behavior) would be the best bet, instead of having an attribute for the behavior?

Reviewer: Yeah, I'm not sure that this is valuable here - are there other places where we handle this? Lots of other controls require focus to be placed back onto the invoking element, but I can't think of a place where we are currently handling that with the primitives themselves. Unless we have a more universal story or principle around handling this, I would probably drop this.

Author: Yeah, even though it is disruptive to refocus after scrolling, this is still a dialog so it should follow that design pattern.

> - `position` - enum, where the tooltip should appear relative to its target, uses Anchored Region logic, see *Position Options*

Author: recapturing original comment: 
https://github.com/microsoft/fast/pull/3691#discussion_r561230159

For more insight: [Managing focus in the shadow DOM](https://nolanlawson.com/2021/02/13/managing-focus-in-the-shadow-dom/)

### Prior Art/Examples

#### Popover from Salesforce
- https://www.lightningdesignsystem.com/components/popovers/

![Popover](./images/salesforce-popover.png)
#### Popover from Evergreen
- triggered by a button
- https://evergreen.segment.com/components/popover/

![Popover](./images/evergreen-popover.png)

#### Microsoft Windows Developer Flyout, Text Only
![Text Only](https://docs.microsoft.com/en-us/windows/uwp/design/controls-and-patterns/images/flyout-wrapping-text.png)

#### Microsoft Windows Developer Flyout, Actionable
![Actionable More Info](https://docs.microsoft.com/en-us/windows/uwp/design/controls-and-patterns/images/flyout-example2.png)

#### Inline Dialog from ATLAssian
- https://atlassian.design/components/inline-dialog/examples

![Popover](./images/atlassian-popover.png)

#### Callout with actions from Fluent UI
- https://developer.microsoft.com/en-us/fluentui#/controls/web/callout
![Callout](./images/fluent-callout.png)

---

### API
*Component Name*
- `fast-flyout`

*Attributes*

> - `visible` - boolean, toggle controlled by the author of whether or not the flyout is visible, defaults to undefined

Reviewer 3: Did we consider logically flipping this attribute and using the platform's 'hidden' attribute instead?

Reviewer 3: I see that Dialog uses 'hidden' and not 'visible', I'd recommend using 'hidden' here both for platform and FAST component alignment.

Author: Godo call

- `trap-focus` - boolean, whether or not to keep focus contained inside the Flyout, defaults to true
- `target` - string, the html id of the HTMLElement that the Flyout is attached to, (viewport for anchored region defaults to the parent of the anchor) and triggered by
- `position` - enum, where the tooltip should appear relative to its target, uses Anchored Region logic, see *Position Options*
- `delay` - number, milliseconds (ms) the time delay before the Flyout is shown once triggered
- `aria-labelledby` - optional based on implementation**
- `aria-describedby` - optional based on implementation**
- `aria-label` - optional, based on implementation**

    ** See the [W3C Specification](https://w3c.github.io/aria-practices/#dialog_roles_states_props) for requirements and details.

*Properties*
- `targetElement` - the reference to the HTMLElement used as target

> - `viewportElement` - the reference to the HTMLElement used as the viewport, required to set up anchored region properly

Reviewer 4: This defaults to the document if unset in anchored region, so viewport can be left unset.  "target" is required.   You could probably have an attribute for viewport as well as target.

Author: Ok I'll swap the required statements.

Author: I rebased which outdated these comments :/ ... I have updated though so let me know what you think!

- `flyoutVisible` - boolean, whether or not the Flyout is visible, for use inside the controller without changing exposed `hidden` attribute

*Slots*
- default for content

*Events*
- `dismiss` - event fired when the Flyout is dismissed

*Functions*
- `createFlyoutTemplate(string: prefix)` - generates a `ViewTemplate` for the Flyout based on the given provided prefix string. This is required as Flyout uses an `anchored-region` internally and the create function generates a template using the appropriate `anchored-region` tag (ie "fast-anchored-region", "fluent-anchored-region"). Note that the appropriate `anchored-region` component must also be declared.

> *Position Options*

Reviewer 4: I'm not clear on the api for positioning.  What choices does the author have for this?   I'd also suggest that authors have the option to either force a position (locktodefault) or be based on available space.

Reviewer 4: Anchored region can do "absolute" or "fixed" positioning.  I think we can elaborate on whether we want to present that choice in the api after we get an initial version in, but what do you see the default being?  "fixed" breaks out of parent containers more easily but absolute doesn't detach from the anchor as often.

Author: I updated the names, does this make more sense? These would be the options available.

I had a `responsive` attribute early on in this spec that would determine if the flyout was forced or would move based on available space, but removed it per some other comments.

I think having something like `responsive` (default to true) which means that the flyout would move along the same axis based upon available space is a great idea. If the flyout `position` option was "right" and `responsive` was true by default then when there was not enough space on the right of the target then the flyout would render on the left of the target. Does that make sense?

Reviewer 4: I'm thinking we should simplify the number of positioning options available.   What if the choices were simply 'top', 'bottom', 'left' and right.  If an author chooses responsive their choice simply flips to the other side based on their choice (ie. left goes to right, top goes to bottom, etc...)?  My thinking is that our base components hava a simple positioning api, and if there is demand we can think about having a "pro" interface where advanced authors could pass a full "anchored region config" object and do all the things.
